includes/embeds/class-amp-youtube-embed-handler.php

```diff
@@ -82,117 +94,313 @@ public function unregister_embed() {
-		$args = wp_parse_args(
-			$args,
-			[
-				'video_id'    => false,
-				'layout'      => 'responsive',
-				'width'       => $this->args['width'],
-				'height'      => $this->args['height'],
-				'placeholder' => '',
-			]
+	public function sanitize_raw_embeds( Document $dom ) {
+
+		$query_segments = array_map(
+			static function ( $domain ) {
+
+				return sprintf(
+					'starts-with( @src, "https://www.%1$s/" ) or starts-with( @src, "https://%1$s/" ) or starts-with( @src, "http://www.%1$s/" ) or starts-with( @src, "http://%1$s/" )',
+					$domain
+				);
+			},
+			$this->applicable_domains
 		);
 
-		if ( empty( $args['video_id'] ) ) {
-			return AMP_HTML_Utils::build_tag(
-				'a',
+		$query = implode( ' or ', $query_segments );
+
+		$nodes = $dom->xpath->query( sprintf( '//iframe[ %s ]', $query ) );
+
+		/** @var Element $node */
+		foreach ( $nodes as $node ) {
+
+			$amp_youtube_component = $this->get_amp_component( $dom, $node );
+
+			if ( ! empty( $amp_youtube_component ) ) {
+				$node->parentNode->replaceChild( $amp_youtube_component, $node );
+			}
+		}
+
+	}
+
+	/**
+	 * Parse YouTube iframe element and return an AMP YouTube component.
+	 *
+	 * @param Document $dom  Document DOM.
+	 * @param Element  $node DOMElement of YouTube iframe.
+	 *
+	 * @return DOMElement|false DOMElement on success, Otherwise false.
+	 */
+	private function get_amp_component( Document $dom, Element $node ) {
+
+		$url      = $node->getAttribute( 'src' );
+		$video_id = $this->get_video_id_from_url( $url );
+
+		if ( ! $video_id ) {
+			return false;
+		}
+
+		$url        = $node->getAttribute( 'src' );
+		$attributes = $this->prepare_attributes( $url );
+
+		if ( ! empty( $node->getAttribute( 'title' ) ) ) {
+			$attributes['title'] = $node->getAttribute( 'title' );
+		}
+
+		if ( ! empty( $node->getAttribute( 'width' ) ) ) {
+			$attributes['width'] = $node->getAttribute( 'width' );
+		}
+
+		if ( ! empty( $node->getAttribute( 'height' ) ) ) {
+			$attributes['height'] = $node->getAttribute( 'height' );
+		}
+
+		if ( empty( $attributes['data-videoid'] ) ) {
+			$link_node = AMP_DOM_Utils::create_node(
+				$dom,
+				Tag::A,
 				[
-					'href'  => esc_url_raw( $url ),
-					'class' => 'amp-wp-embed-fallback',
-				],
-				esc_html( $url )
+					Attribute::HREF   => esc_url_raw( $url ),
+					Attribute::CLASS_ => 'amp-wp-embed-fallback',
+				]
 			);
+			$link_node->appendChild( $dom->createTextNode( $url ) );
+
+			return $link_node;
 		}
 
-		$this->did_convert_elements = true;
+		$video_title = ( ! empty( $attributes['title'] ) ) ? $attributes['title'] : null;
 
-		$attributes = array_merge(
-			[ 'data-videoid' => $args['video_id'] ],
-			wp_array_slice_assoc( $args, [ 'layout', 'width', 'height' ] )
+		$amp_node = AMP_DOM_Utils::create_node(
+			$dom,
+			'amp-youtube',
+			$attributes
 		);
-		if ( ! empty( $args['title'] ) ) {
-			$attributes['title'] = $args['title'];
+
+		if ( ! empty( $amp_node ) && is_a( $amp_node, 'DOMElement' ) ) {
```

Comment:
Best to specify the full namespace of the class as to avoid any issues if this class were to be `namespace`d.

```suggestion
		if ( ! empty( $amp_node ) && is_a( $amp_node, '\DOMElement' ) ) {
```